File: Analyzer.cpp
#include <iostream>
#include <string>
#include <vector>
#include "StringData.h"


using namespace std;

int linearSearch(string element, vector<string> dataSet){
    for(int i = 0; i < dataSet.size(); i++)
    {
        if((dataSet[i])==(element)){
            return i;
        }
    }
    return -1;
}

int binarySearch(string element, vector<string> dataset){
    int minIndex = 0;
    int maxIndex = dataset.size()-1;
    int mid = 0;

    while (minIndex <= maxIndex){
        mid = dataset.size()/2;
        if(dataset[mid].compare(element)==0)
        {
            return mid;
        }
        else if(dataset[mid].compare(element)<0)
        {
            minIndex = mid + 1;
        }
        else
        {
            maxIndex = mid - 1;
        }

    }
    return -1;
}



int main() {
    //std::cout << "Hello, World!" << std::endl;
    vector <string> dataset = getStringData();

    string element = "";

    auto start = 0;
    auto end = 0;
    int outdex = 0;
    string printOut = "";

    element = "not_here";
    start = systemTimeNanoseconds();
    outdex = linearSearch(element, dataset);
    end = systemTimeNanoseconds();
    printOut = "Linear Search: " + to_string(outdex) + " Time: " + to_string(end-start-326);
    cout << printOut << "\n";

    start = systemTimeNanoseconds();
    outdex = binarySearch(element, dataset);
    end = systemTimeNanoseconds();
    printOut = "Binary Search: " + to_string(outdex) + " Time: " + to_string(end-start-326);
    cout << printOut<<"\n";

    element = "aaaaa";
    start = systemTimeNanoseconds();
    outdex = linearSearch(element, dataset);
    end = systemTimeNanoseconds();
    printOut = "Linear Search: " + to_string(outdex) + " Time: " + to_string(end-start-326);
    cout << printOut<<"\n";

    start = systemTimeNanoseconds();
    outdex = binarySearch(element, dataset);
    end = systemTimeNanoseconds();
    printOut = "Binary Search: " + to_string(outdex) + " Time: " + to_string(end-start-326);
    cout << printOut<<"\n";

    element = "mzzzz";
    start = systemTimeNanoseconds();
    outdex = linearSearch(element, dataset);
    end = systemTimeNanoseconds();
    printOut = "Linear Search: " + to_string(outdex) + " Time: " + to_string(end-start-326);
    cout << printOut<<"\n";

    start = systemTimeNanoseconds();
    outdex = binarySearch(element, dataset);
    end = systemTimeNanoseconds();
    printOut = "Binary Search: " + to_string(outdex) + " Time: " + to_string(end-start-326);
    cout << printOut<<"\n";


    return 0;


}

File: Analyzer.java
import java.lang.String;
public class Analyzer
{
    public static int linearSearch(String[] dataSet, String elements)
    {
        for(int i = 0; i < dataSet.length; i++)
        {
            if(dataSet[i].equals(elements))
            {
                return i;
            }
        }
        return -1;
    }
    public static int binarySearch(String[] dataSet, String element)
    {

        int index = dataSet.length/2;

        int min = 0;
        int max = dataSet.length-1;

        int count = 0;

        while(min<=max)
        {
            if(dataSet[index].compareTo(element)==0)
            {
                return index;
            }
            else if(dataSet[index].compareTo(element)<0)
            {
                min = index+1;
            }
            else
            {
                max = index-1;
            }
            index=(max+min)/2;
            count++;
        }
        return -1;
    }


    public static void main(String[] args)
    {
        String[] dataSet = StringData.getData();

        double timeBefore;
        double timeAfter;

        int index;

        timeBefore= System.nanoTime();
        index = linearSearch(dataSet, "not_here");
        timeAfter= System.nanoTime();
        System.out.println("Linear Search: " + index + " Time: " + (timeAfter-timeBefore));

        timeBefore= System.nanoTime();
        index = binarySearch(dataSet, "not_here");
        timeAfter= System.nanoTime();
        System.out.println("Binary Search: "  + index + " Time: " + (timeAfter-timeBefore));

        timeBefore= System.nanoTime();
        index = linearSearch(dataSet, "mzzzz");
        timeAfter= System.nanoTime();
        System.out.println("Linear Search: " + index + " Time: " + (timeAfter-timeBefore));

        timeBefore= System.nanoTime();
        index = binarySearch(dataSet, "mzzzz");
        timeAfter= System.nanoTime();
        System.out.println("Binary Search: "  + index + " Time: " + (timeAfter-timeBefore));

        timeBefore= System.nanoTime();
        index = linearSearch(dataSet, "aaaaa");
        timeAfter= System.nanoTime();
        System.out.println("Linear Search: " + index + " Time: " + (timeAfter-timeBefore));

        timeBefore= System.nanoTime();
        index = binarySearch(dataSet, "aaaaa");
        timeAfter= System.nanoTime();
        System.out.println("Binary Search: "  + index + " Time: " + (timeAfter-timeBefore));



    }

}
